File: plugins/cc-bridge/src/readers/hooks.ts
import fs from "node:fs";
import path from "node:path";

export type HookHandler = { type: string; [key: string]: unknown };
export type HookEntry = { matcher: string; hooks: HookHandler[] };
export type HooksMap = Record<string, HookEntry[]>;

export function readHooks(projectRoot: string): HooksMap {
  return loadHooks(path.join(projectRoot, ".claude", "settings.json"));
}

function loadHooks(settingsPath: string): HooksMap {
  if (!isFile(settingsPath)) return {};
  let raw: string;
  try {
    raw = fs.readFileSync(settingsPath, "utf-8");
  } catch {
    return {};
  }
  let data: unknown;
  try {
    data = JSON.parse(raw);
  } catch {
    return {};
  }
  if (!isPlainObject(data)) return {};
  const hooks = (data as Record<string, unknown>).hooks;
  if (!isPlainObject(hooks)) return {};
  return hooks as HooksMap;
}

function isFile(p: string): boolean {
  try {
    return fs.statSync(p).isFile();
  } catch {
    return false;
  }
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: plugins/cc-bridge/scripts/readers/hooks.py
"""Reader for Claude Code hooks configuration.

Reads the ``hooks`` key from Claude Code ``settings.json`` files at both the
project and user scopes and merges them. Project-scope events override
user-scope events with the same event name; user-only events are kept.
"""

from __future__ import annotations

import json
from pathlib import Path


def _load_hooks(settings_path: Path) -> dict:
    """Return the ``hooks`` dict from a settings.json, or {} if unavailable."""
    if not settings_path.is_file():
        return {}
    try:
        data = json.loads(settings_path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return {}
    if not isinstance(data, dict):
        return {}
    hooks = data.get("hooks")
    if not isinstance(hooks, dict):
        return {}
    return hooks


def read_hooks(project_root: Path, user_home: Path | None = None) -> dict:
    """Read and merge Claude Code hooks from project and user settings.

    Project events take precedence over user events with the same event name.
    Events only present in user settings are preserved.
    """
    project_hooks = _load_hooks(Path(project_root) / ".claude" / "settings.json")
    user_hooks: dict = {}
    if user_home is not None:
        user_hooks = _load_hooks(Path(user_home) / ".claude" / "settings.json")

    merged: dict = {}
    merged.update(user_hooks)
    merged.update(project_hooks)
    return merged
